### PR TITLE
Set prompt loss to zero

### PR DESCRIPTION
**Pull request by l-k-11235**
This PR allows to set the prompt loss to zero when finetuning an LM on a specific task.

### CODE REVIEW

File: onmt/utils/loss.py [hunk 1]
```diff
@@ -261,6 +261,51 @@ def _bottle(self, _v):
     def _unbottle(self, _v, batch_size):
         return _v.view(-1, batch_size, _v.size(1))
 
+    def ignore_prompt(
+        self, batch, response_token="▁Response", response_right_pattern_length=3
+    ):
+        """
+        Mask the prompt in the target side of the bath examples in order
```
Comment (francoishernandez) [inline onmt/utils/loss.py:268]: Typo "batch"

File: onmt/utils/loss.py [hunk 2]
```diff
@@ -261,6 +261,51 @@ def _bottle(self, _v):
     def _unbottle(self, _v, batch_size):
         return _v.view(-1, batch_size, _v.size(1))
 
+    def ignore_prompt(
+        self, batch, response_token="▁Response", response_right_pattern_length=3
```
Comment (francoishernandez) [inline onmt/utils/loss.py:265]: Do we want a more specific placeholder, e.g. `｟_response_｠`. This "▁Response" feels too generic and we might have some cases where we'd unwillingly mask some content that should not be masked.

File: onmt/utils/loss.py [hunk 3]
```diff
@@ -261,6 +261,51 @@ def _bottle(self, _v):
...
+            to set the loss of the prompt to zero.
+        For finetuning on specific tasks.
+        The masks are supposed to be properly handled by the loss criterion
+            (e.g. nn.CrossEntropyLoss )
+
+        Args:
+        batch: The current batch.
+        response_token (str): The token used the locate the end of the prompt
+        response_right_pattern_length (int). The number of tokens in the response
+            pattern after the response_token
```
Comment (francoishernandez) [inline onmt/utils/loss.py:278]: This does not look like proper syntax.

File: onmt/utils/loss.py [hunk 4]
```diff
@@ -261,6 +261,51 @@ def _bottle(self, _v):
...
+
+        Ex: The response pattern is '### Response : ｟newline｠',
+            the response_token is '▁Response'.
+            and the tokenized response pattern is '##', '#', '▁Response', '▁:', '▁', '<0x0A>'].
+            Then response_right_pattern_length = 3
```
Comment (francoishernandez) [inline onmt/utils/loss.py:283]: Should probably be before the Args section for proper doc syntax.

File: onmt/utils/loss.py (hunk 2 above)
Comment (francoishernandez) [reply onmt/utils/loss.py:265]: Or at least make it configurable via an opt flag.

File: onmt/utils/loss.py [hunk 5]
```diff
@@ -261,6 +261,51 @@ def _bottle(self, _v):
...
 
+    def ignore_prompt(
+        self, batch, response_token="▁Response", response_right_pattern_length=3
+    ):
+        """
+        Mask the prompt in the target side of the bath examples in order
+            to set the loss of the prompt to zero.
+        For finetuning on specific tasks.
+        The masks are supposed to be properly handled by the loss criterion
+            (e.g. nn.CrossEntropyLoss )
```
Comment (francoishernandez) [inline onmt/utils/loss.py:272]: Do we want to add some kind of reference around here? (Ghost attention, other ?)

File: onmt/utils/loss.py [hunk 6]
```diff
@@ -261,6 +261,51 @@ def _bottle(self, _v):
...
+
+        Ex: The response pattern is '### Response : ｟newline｠',
+            the response_token is '▁Response'.
+            and the tokenized response pattern is '##', '#', '▁Response', '▁:', '▁', '<0x0A>'].
+            Then response_right_pattern_length = 3
+        """
+        import numpy as np
+
+        response_idx = self.vocab[response_token]  # 13291
+        ignore_index = self.padding_idx
```
Comment (francoishernandez) [inline onmt/utils/loss.py:288]: Feels weird to have idx and index, we should probably homogenize to idx

File: onmt/utils/loss.py [hunk 7]
```diff
@@ -261,6 +261,51 @@ def _bottle(self, _v):
...
+        response_token (str): The token used the locate the end of the prompt
+        response_right_pattern_length (int). The number of tokens in the response
+            pattern after the response_token
+
+        Ex: The response pattern is '### Response : ｟newline｠',
+            the response_token is '▁Response'.
+            and the tokenized response pattern is '##', '#', '▁Response', '▁:', '▁', '<0x0A>'].
+            Then response_right_pattern_length = 3
+        """
+        import numpy as np
```
Comment (francoishernandez) [inline onmt/utils/loss.py:285]: Can't we make it work with torch since it's already imported and used for the rest?

File: onmt/utils/loss.py [hunk 8]
```diff
@@ -261,6 +261,51 @@ def _bottle(self, _v):
...
+
+        response_idx = self.vocab[response_token]  # 13291
+        ignore_index = self.padding_idx
+        nb_examples = batch["src"].size()[0]
+        for j in range(nb_examples):
+            # Locate the end of the prompt
+            response_start = np.where(
+                batch["src"][j, :, 0].cpu().numpy() == response_idx
+            )[0].tolist()[0]
+            response_start += response_right_pattern_length
```
Comment (francoishernandez) [inline onmt/utils/loss.py:295]: This `response_right_pattern_length` trick feels a bit weird to me (and probably hard to generalize, considering you would need to know how it's tokenized, etc.
Can't we make it work with just using some tokens? E.g. adding some kind of placeholder `｟_mask_before_｠` and put it where we would not need such trick.

File: onmt/trainer.py [hunk 9]
```diff
@@ -496,14 +496,15 @@ def _gradient_accumulation(
                         bptt = True
 
                         # 3. Compute loss.
+                        # The loss of the prompt will be set to zero.
+                        batch = self.train_loss.ignore_prompt(batch)
```
Comment (francoishernandez) [inline onmt/trainer.py:500]: We might want to enable/disable this with some dedicated opt flag. (It should be transparent in theory, but probably not useful to make it run when not needed.)

**Review by francoishernandez — COMMENTED** [review]
Cool first implementation. Few comments below to try and make it a bit more general.

File: onmt/utils/loss.py (hunk 5 above)
Comment (l-k-11235) [reply onmt/utils/loss.py:272]: Ghost attention is something else.
It is mentioned is various papers, like
- The Self-instruct paper (alpaca finetuning)
image
- The Llama-2 paper:
image

Comment (francoishernandez) [reply onmt/utils/loss.py:272]: The Llama2 paper is confusing. This 0 loss thing seems to be part of the Ghost Attention method, but indeed there might be other aspects to this.
Screenshot from 2023-09-04 16-31-03

File: onmt/transforms/tokenize.py [hunk 10]
```diff
@@ -208,59 +249,43 @@ def warm_up(self, vocabs=None):
...
+            string,
+            out_type=str,
+            enable_sampling=True,
+            alpha=alpha,
+            nbest_size=nbest_size,
+        )
+        return tokens
 
-        return segmented[:-1]
+    def segment_determistic(self, string, side="src"):
```
Comment (francoishernandez) [inline onmt/transforms/tokenize.py:266]: typo "deterministic" ?

File: onmt/transforms/tokenize.py [hunk 11]
```diff
@@ -208,59 +249,43 @@ def warm_up(self, vocabs=None):
...
+        print("not sampling ...")
```
Comment (francoishernandez) [inline onmt/transforms/tokenize.py:267]: Not sure we want to keep this print.

File: onmt/transforms/tokenize.py [hunk 12]
```diff
@@ -208,59 +249,43 @@ def warm_up(self, vocabs=None):
...
+        sp_model = self.load_models[side]
+        tokens = sp_model.encode(string, out_type=str)
+        return tokens
+
+    def tokenize_string(self, string, side="src", is_train=False):
+        """Apply subword sampling or deterministic subwording"""
+        nbest_size = self.tgt_subword_nbest if side == "tgt" else self.src_subword_nbest
+        if is_train is False or nbest_size in [0, 1]:
+            # derterministic subwording
+            tokens = self.segment_determistic(string, side)
```
Comment (francoishernandez) [inline onmt/transforms/tokenize.py:277]: Maybe we should keep "tokenize_xx" instead of "segment_xx", as segment could be confusing with "sentence/paragraph" segmentation rather than the segmentation of tokens here.

**Review by francoishernandez — COMMENTED** [review]
Looks good! Do we have unit tests already covering this or do we maybe want to add a few simple ones while we're at it?
